sdk/python/feast/infra/registry/caching_registry.py

```diff
@@ -289,22 +291,36 @@ def refresh(self, project: Optional[str] = None):
         self.cached_registry_proto_created = datetime.utcnow()
 
     def _refresh_cached_registry_if_necessary(self):
-        with self._refresh_lock:
-            expired = (
-                self.cached_registry_proto is None
-                or self.cached_registry_proto_created is None
-            ) or (
-                self.cached_registry_proto_ttl.total_seconds()
-                > 0  # 0 ttl means infinity
-                and (
-                    datetime.utcnow()
-                    > (
-                        self.cached_registry_proto_created
-                        + self.cached_registry_proto_ttl
+        if self.cache_mode == "sync":
+            with self._refresh_lock:
+                expired = (
+                    self.cached_registry_proto is None
+                    or self.cached_registry_proto_created is None
+                ) or (
+                    self.cached_registry_proto_ttl.total_seconds()
+                    > 0  # 0 ttl means infinity
+                    and (
+                        datetime.utcnow()
+                        > (
+                            self.cached_registry_proto_created
+                            + self.cached_registry_proto_ttl
+                        )
                     )
                 )
-            )
 
-            if expired:
-                logger.info("Registry cache expired, so refreshing")
-                self.refresh()
+                if expired:
+                    logger.info("Registry cache expired, so refreshing")
+                    self.refresh()
+
+    def _start_thread_async_refresh(self, cache_ttl_seconds):
+        self.refresh()
+        if cache_ttl_seconds <= 0:
+            return
+        self.registry_refresh_thread = threading.Timer(
```

Comment:
Assume this creates new threads for each call in recursion? Can you check and confirm once?

Reply:
Sorry. Not able to reproduce this.

Reply:
@EXPEbdodla The way I read it, each new thread will 1) refresh 2) spawn off another timer thread (creating an infinite loop) and exit. Am I missing something?

Reply:
That's right. I was under impression we need to explicitly cancel/stop the spawned thread. Timer class is doing that automatically.

Reply:
Yeah, this is also how it's currently implemented in python feature server. Probably not the best way though as each iteration creates a brand new os-level thread. We might have to revisit that later.